File: app-macos/Tests/RunnerForgeTests/TemplateRendererTests.swift
import Foundation
import Testing
@testable import RunnerForge

/// A workflow that still contains a {{placeholder}} is accepted by git and
/// rejected by Actions, minutes later and far less clearly. The renderer's job
/// is to make that impossible.
@Suite("TemplateRenderer")
struct TemplateRendererTests {

    @Test("substitutes every occurrence of a key")
    func substitutesRepeatedKeys() throws {
        let output = try TemplateRenderer().render(
            "{{name}} and {{name}} again", values: ["name": "Canary"])
        #expect(output == "Canary and Canary again")
    }

    @Test("a leftover placeholder is an error, not a warning")
    func leftoverPlaceholderThrows() {
        #expect(throws: TemplateRenderer.UnsubstitutedPlaceholderError.self) {
            try TemplateRenderer().render("runs-on: {{missing}}", values: [:])
        }
    }

    /// The error has to say WHICH placeholder, or the user is left diffing a
    /// 300-line workflow by eye.
    @Test("the error names every missing key")
    func errorNamesMissingKeys() {
        do {
            _ = try TemplateRenderer().render("{{alpha}} {{beta}}", values: ["alpha": "a"])
            Issue.record("expected the render to throw")
        } catch let error as TemplateRenderer.UnsubstitutedPlaceholderError {
            #expect(error.keys == ["beta"])
            #expect(error.description.contains("beta"))
        } catch {
            Issue.record("wrong error type: \(error)")
        }
    }

    @Test("placeholders(in:) lists keys once, sorted")
    func placeholderDiscovery() {
        let found = TemplateRenderer().placeholders(in: "{{b}} {{a}} {{b}}")
        #expect(found == ["a", "b"])
    }

    @Test("a template with no placeholders is returned unchanged")
    func noPlaceholders() throws {
        let text = "name: build\non: push\n"
        #expect(try TemplateRenderer().render(text, values: [:]) == text)
    }

    /// The substitution map must cover the shipped templates. This is the test
    /// that catches a template gaining a placeholder nobody wired up.
    @Test("buildValues covers every placeholder in the shipped templates")
    func buildValuesCoversTemplates() throws {
        let repoRoot = URL(fileURLWithPath: #filePath)
            .deletingLastPathComponent()   // RunnerForgeTests
            .deletingLastPathComponent()   // Tests
            .deletingLastPathComponent()   // app-macos
            .deletingLastPathComponent()   // repo root
        let templates = repoRoot.appendingPathComponent("templates")

        guard FileManager.default.fileExists(atPath: templates.path) else {
            // Running from an installed copy rather than the repo. Say so rather
            // than passing silently on an assertion that never ran.
            Issue.record("templates/ not found at \(templates.path); this test must run from the repo")
            return
        }

        var config = ForgeConfig.makeDefault(workDir: "/tmp/rf")
        config.github = GitHubConfig(
            owner: "example", repos: ["plugin"], appId: "1", installationId: "2")

        let values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".",
            secretPresence: Dictionary(uniqueKeysWithValues: SecretStore.keyNames.map { ($0, true) }))

        let renderer = TemplateRenderer()
        let names = try FileManager.default.contentsOfDirectory(atPath: templates.path)
            .filter { $0.hasSuffix(".tmpl") }
        #expect(!names.isEmpty)

        for name in names {
            let text = try String(
                contentsOf: templates.appendingPathComponent(name), encoding: .utf8)
            let missing = renderer.placeholders(in: text).filter { values[$0] == nil }
            #expect(missing.isEmpty, "\(name) has unmapped placeholders: \(missing)")
        }
    }

    /// --allowsigningservice belongs to cloud mode only. Emitting it with a
    /// physical dongle attached is a wraptool error, not a harmless extra flag.
    @Test("the AAX requirement flags follow the SDK setting")
    func aaxFlagsFollowSdk() {
        var config = ForgeConfig.makeDefault(workDir: "/tmp/rf")

        config.paths.aaxSdkSource = ""
        var values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".", secretPresence: [:])
        #expect(values["aaxRequired"] == "false")
        #expect(values["aaxSkipRequired"] == "true")

        config.paths.aaxSdkSource = "/Volumes/sdk/aax.zip"
        values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".", secretPresence: [:])
        #expect(values["aaxRequired"] == "true")
        #expect(values["aaxSkipRequired"] == "false")
    }

    /// Secret PRESENCE reaches the templates; secret VALUES never do.
    @Test("only presence, never a value, reaches the substitution map")
    func secretsNeverReachTemplates() {
        let config = ForgeConfig.makeDefault(workDir: "/tmp/rf")
        let values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".",
            secretPresence: ["pacePassword": true, "appleDevIdP12": false])

        #expect(values["havePacePassword"] == "yes")
        #expect(values["haveAppleP12"] == "no")
        for value in values.values {
            #expect(!value.contains("BEGIN PRIVATE KEY"))
            #expect(!value.contains("BEGIN RSA PRIVATE KEY"))
        }
    }

    // -----------------------------------------------------------------------
    // Publisher arguments
    //
    // wraptool takes EITHER a Wrap Config GUID OR a customer number paired with
    // the company name, and rejects the number on its own. These tests are the
    // only place that rule is checked before a twenty-minute build spends itself
    // proving it.
    // -----------------------------------------------------------------------

    private func exampleConfig() -> ForgeConfig {
        var config = ForgeConfig.makeDefault(workDir: "/tmp/rf")
        config.github = GitHubConfig(
            owner: "example", repos: ["plugin"], appId: "1", installationId: "2")
        return config
    }

    @Test("a Wrap Config GUID becomes a single wcguid argument")
    func wrapConfigBecomesOneArgument() {
        var config = exampleConfig()
        config.signing.paceWcGuid = "ED052BE6-9B79-4E72-A838-E226320B778E"

        #expect(TemplateRenderer.paceIdentityArgs(config, shell: true)
                == "--wcguid 'ED052BE6-9B79-4E72-A838-E226320B778E'")
        #expect(TemplateRenderer.paceIdentityArgs(config, shell: false)
                == "-WcGuid 'ED052BE6-9B79-4E72-A838-E226320B778E'")
    }

    @Test("a customer number always carries the company name with it")
    func customerNumberCarriesName() {
        var config = exampleConfig()
        config.signing.paceWcGuid = ""
        config.signing.paceCustomerNumber = "ABCD-1234-ABCD-1234"
        config.signing.paceCustomerName = "My Company, Inc."

        let shell = TemplateRenderer.paceIdentityArgs(config, shell: true)
        #expect(shell.contains("--customernumber 'ABCD-1234-ABCD-1234'"))
        #expect(shell.contains("--customername 'My Company, Inc.'"))
        #expect(!shell.contains("--wcguid"))

        let pwsh = TemplateRenderer.paceIdentityArgs(config, shell: false)
        #expect(pwsh.contains("-CustomerNumber 'ABCD-1234-ABCD-1234'"))
        #expect(pwsh.contains("-CustomerName 'My Company, Inc.'"))
    }

    /// The failure this guards against: emitting a workflow with no publisher
    /// argument at all, which fails inside wraptool minutes later with a message
    /// that never mentions the configuration.
    @Test("a customer number without a name emits a marker, never an empty string")
    func incompleteCustomerEmitsMarker() {
        var config = exampleConfig()
        config.signing.paceWcGuid = ""
        config.signing.paceCustomerNumber = "ABCD-1234-ABCD-1234"
        config.signing.paceCustomerName = ""

        let shell = TemplateRenderer.paceIdentityArgs(config, shell: true)
        #expect(!shell.trimmingCharacters(in: .whitespaces).isEmpty)
        #expect(shell.contains("SET-A-WRAP-CONFIG"))
        #expect(!shell.contains("ABCD-1234"))
    }

    @Test("publisher identity needs a GUID, or both customer fields")
    func publisherIdentityRule() {
        var config = exampleConfig()

        config.signing.paceWcGuid = "G"
        config.signing.paceCustomerNumber = ""
        config.signing.paceCustomerName = ""
        #expect(config.signing.hasPublisherIdentity)

        config.signing.paceWcGuid = ""
        config.signing.paceCustomerNumber = "N"
        #expect(!config.signing.hasPublisherIdentity)

        config.signing.paceCustomerName = "Co"
        #expect(config.signing.hasPublisherIdentity)
    }

    @Test("the self-signed flag is emitted only when the certificate is self-signed")
    func selfSignedFlag() {
        var config = exampleConfig()

        config.signing.paceSelfSigned = false
        var values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".", secretPresence: [:])
        #expect(values["paceSelfSignedSh"] == "")
        #expect(values["paceSelfSignedPs"] == "")

        config.signing.paceSelfSigned = true
        values = TemplateRenderer.buildValues(
            config: config, projectName: "Canary", sourceDir: ".", secretPresence: [:])
        #expect(values["paceSelfSignedSh"]?.contains("--self-signed") == true)
        #expect(values["paceSelfSignedPs"]?.contains("-SelfSigned") == true)
    }

    /// A forge.json written before these fields existed must still load. A
    /// decoder that throws on a missing key turns an upgrade into a crash.
    @Test("a signing block without the new fields still decodes")
    func olderConfigStillDecodes() throws {
        let json = """
        {"mode":"windows-ilok","paceAccount":"acct","paceWcGuid":"G","paceSignId":"S",
         "allowSigningService":true,
         "windows":{"provider":"none","azureEndpoint":"","azureAccount":"","azureProfile":""},
         "macos":{"teamId":"","devIdAppIdentity":"","devIdInstallerIdentity":"","notarize":true}}
        """
        let signing = try JSONDecoder().decode(SigningConfig.self, from: Data(json.utf8))

        #expect(signing.paceWcGuid == "G")
        #expect(signing.paceCustomerNumber == "")
        #expect(signing.paceCustomerName == "")
        #expect(signing.paceSelfSigned == false)
        #expect(signing.hasPublisherIdentity)
    }
}

File: app-windows/RunnerForge.Tests/TemplateRendererTests.cs
using RunnerForge.Models;
using RunnerForge.Services;
using Xunit;

namespace RunnerForge.Tests;

/// <summary>
/// The renderer is what turns configuration into the YAML a user pastes into
/// their repository, so its failure mode matters: an unsubstituted placeholder
/// must be an ERROR, never something quietly emitted.
/// </summary>
public sealed class TemplateRendererTests
{
    private readonly TemplateRenderer _renderer = new();

    [Fact]
    public void Substitutes_every_known_placeholder()
    {
        string result = _renderer.Render(
            "runs-on: {{labels}}\nname: {{project}}",
            new Dictionary<string, string> { ["labels"] = "[a,b]", ["project"] = "Canary" });

        Assert.Equal("runs-on: [a,b]\nname: Canary", result);
    }

    [Fact]
    public void Substitutes_a_repeated_placeholder_everywhere()
    {
        string result = _renderer.Render(
            "{{p}}-windows-x64 {{p}}-macos-universal {{p}}-logs",
            new Dictionary<string, string> { ["p"] = "Canary" });

        Assert.Equal("Canary-windows-x64 Canary-macos-universal Canary-logs", result);
    }

    /// <summary>
    /// Emitting YAML that still contains {{something}} would produce a workflow
    /// that fails at run time on the user's machine. Far better to refuse.
    /// </summary>
    [Fact]
    public void An_unsubstituted_placeholder_throws_rather_than_being_emitted()
    {
        TemplateRenderer.UnsubstitutedPlaceholderException exception =
            Assert.Throws<TemplateRenderer.UnsubstitutedPlaceholderException>(
                () => _renderer.Render("a: {{known}} b: {{missing}}",
                    new Dictionary<string, string> { ["known"] = "1" }));

        Assert.Contains("missing", exception.Keys);
        Assert.DoesNotContain("known", exception.Keys);
    }

    [Fact]
    public void Reports_every_missing_placeholder_not_just_the_first()
    {
        TemplateRenderer.UnsubstitutedPlaceholderException exception =
            Assert.Throws<TemplateRenderer.UnsubstitutedPlaceholderException>(
                () => _renderer.Render("{{a}} {{b}} {{c}}", new Dictionary<string, string>()));

        Assert.Equal(["a", "b", "c"], exception.Keys);
    }

    [Fact]
    public void Text_that_merely_looks_like_a_placeholder_is_left_alone()
    {
        // Shell parameter expansion and GitHub expressions both use braces, and
        // neither may be mangled.
        const string template = "run: echo ${HOME} and ${{ github.sha }}";
        string result = _renderer.Render(template, new Dictionary<string, string>());
        Assert.Equal(template, result);
    }

    [Fact]
    public void Lists_the_placeholders_a_template_uses()
    {
        IReadOnlyList<string> found = _renderer.PlaceholdersIn("{{b}} {{a}} {{b}}");
        Assert.Equal(["a", "b"], found);
    }

    [Fact]
    public void Built_values_render_label_sets_as_json_arrays()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        Dictionary<string, string> values =
            TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());

        Assert.Equal("""["self-hosted","windows","x64","container","forge"]""", values["winBuildLabelsJson"]);
        Assert.Equal("""["self-hosted","macos","arm64","tart","forge"]""", values["macBuildLabelsJson"]);
        Assert.Equal("""["self-hosted","windows","x64","ilok","forge"]""", values["winIlokLabelsJson"]);
    }

    /// <summary>
    /// With no AAX SDK the AAX artifacts become optional and the skip marker
    /// becomes required, so the artifact set is never silently short.
    /// </summary>
    [Fact]
    public void Without_an_aax_sdk_the_skip_marker_becomes_required()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Paths.AaxSdkSource = "";

        Dictionary<string, string> values =
            TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());

        Assert.Equal("false", values["aaxRequired"]);
        Assert.Equal("true", values["aaxSkipRequired"]);
    }

    [Fact]
    public void With_an_aax_sdk_the_aax_artifacts_become_required()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Paths.AaxSdkSource = @"C:\sdk\aax.zip";

        Dictionary<string, string> values =
            TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());

        Assert.Equal("true", values["aaxRequired"]);
        Assert.Equal("false", values["aaxSkipRequired"]);
    }

    [Fact]
    public void Secret_presence_is_reported_as_yes_or_no_and_never_as_a_value()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");

        Dictionary<string, string> values = TemplateRenderer.BuildValues(
            config, "Canary", ".",
            new Dictionary<string, bool> { ["pacePassword"] = true, ["azureClientId"] = false });

        Assert.Equal("yes", values["havePacePassword"]);
        Assert.Equal("no", values["haveAzureClientId"]);
    }

    /// <summary>
    /// Setting the provider to none must disable the signing step rather than
    /// emitting a step that will fail for want of credentials.
    /// </summary>
    [Fact]
    public void Windows_signing_is_disabled_when_the_provider_is_none()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Signing.Windows.Provider = "none";

        Dictionary<string, string> values =
            TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());

        Assert.Equal("false", values["windowsSigningEnabled"]);
    }

    // -----------------------------------------------------------------------
    // Publisher arguments
    //
    // wraptool takes EITHER a Wrap Config GUID OR a customer number paired with
    // the company name, and rejects the number on its own. These four tests are
    // the only place that rule is checked before a twenty-minute build spends
    // itself proving it.
    // -----------------------------------------------------------------------

    [Fact]
    public void A_wrap_config_guid_becomes_a_single_wcguid_argument()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Signing.PaceWcGuid = "ED052BE6-9B79-4E72-A838-E226320B778E";

        Assert.Equal("--wcguid 'ED052BE6-9B79-4E72-A838-E226320B778E'",
                     TemplateRenderer.PaceIdentityArgs(config, shell: true));
        Assert.Equal("-WcGuid 'ED052BE6-9B79-4E72-A838-E226320B778E'",
                     TemplateRenderer.PaceIdentityArgs(config, shell: false));
    }

    [Fact]
    public void A_customer_number_always_carries_the_company_name_with_it()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Signing.PaceWcGuid = "";
        config.Signing.PaceCustomerNumber = "ABCD-1234-ABCD-1234";
        config.Signing.PaceCustomerName = "My Company, Inc.";

        string shell = TemplateRenderer.PaceIdentityArgs(config, shell: true);
        Assert.Contains("--customernumber 'ABCD-1234-ABCD-1234'", shell);
        Assert.Contains("--customername 'My Company, Inc.'", shell);
        Assert.DoesNotContain("--wcguid", shell);

        string pwsh = TemplateRenderer.PaceIdentityArgs(config, shell: false);
        Assert.Contains("-CustomerNumber 'ABCD-1234-ABCD-1234'", pwsh);
        Assert.Contains("-CustomerName 'My Company, Inc.'", pwsh);
    }

    /// <summary>
    /// The failure mode this guards against: emitting a workflow with no
    /// publisher argument at all, which fails inside wraptool minutes later with
    /// a message that never mentions the configuration.
    /// </summary>
    [Fact]
    public void A_customer_number_without_a_name_emits_a_marker_not_an_empty_string()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Signing.PaceWcGuid = "";
        config.Signing.PaceCustomerNumber = "ABCD-1234-ABCD-1234";
        config.Signing.PaceCustomerName = "";

        string shell = TemplateRenderer.PaceIdentityArgs(config, shell: true);
        Assert.NotEqual("", shell.Trim());
        Assert.Contains("SET-A-WRAP-CONFIG", shell);
        Assert.DoesNotContain("ABCD-1234", shell);
    }

    [Fact]
    public void Nothing_configured_at_all_also_emits_the_marker()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");
        config.Signing.PaceWcGuid = "";
        config.Signing.PaceCustomerNumber = "";
        config.Signing.PaceCustomerName = "";

        Assert.Contains("SET-A-WRAP-CONFIG", TemplateRenderer.PaceIdentityArgs(config, shell: true));
        Assert.False(config.Signing.HasPublisherIdentity);
    }

    [Fact]
    public void Publisher_identity_needs_a_guid_or_both_customer_fields()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");

        config.Signing.PaceWcGuid = "G";
        config.Signing.PaceCustomerNumber = "";
        config.Signing.PaceCustomerName = "";
        Assert.True(config.Signing.HasPublisherIdentity);

        config.Signing.PaceWcGuid = "";
        config.Signing.PaceCustomerNumber = "N";
        Assert.False(config.Signing.HasPublisherIdentity);

        config.Signing.PaceCustomerName = "Co";
        Assert.True(config.Signing.HasPublisherIdentity);
    }

    [Fact]
    public void The_self_signed_flag_is_emitted_only_when_the_certificate_is_self_signed()
    {
        ForgeConfig config = ForgeConfig.CreateDefault("/tmp/work");

        config.Signing.PaceSelfSigned = false;
        var off = TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());
        Assert.Equal("", off["paceSelfSignedSh"]);
        Assert.Equal("", off["paceSelfSignedPs"]);

        config.Signing.PaceSelfSigned = true;
        var on = TemplateRenderer.BuildValues(config, "Canary", ".", new Dictionary<string, bool>());
        Assert.Contains("--self-signed", on["paceSelfSignedSh"]);
        Assert.Contains("-SelfSigned", on["paceSelfSignedPs"]);
    }
}
